reco_utils/recommender/fbt/fbt.py

```diff
@@ -0,0 +1,302 @@
+# Copyright (c) Microsoft Corporation. All rights reserved.
+# Licensed under the MIT License.
+
+from reco_utils.common import constants
+from reco_utils.evaluation.python_evaluation import get_top_k_items
+import pandas as pd
+import logging
+
+logger = logging.getLogger()
+
+
+class FBT(object):
+    """Frequently Bought Together Algorithm for Recommendations
+    (FBT) implementation
+
+    FBT is a fast scalable adaptive algorithm for personalized
+    recommendations based on user-item interaction history.
+    The core idea behind FBT: Given that a user has interacted
+    with an item, what other items were most frequently interacted with
+    by other users? Lets recommend these other items to our current user!
+    """
+
+    def __init__(
+        self,
+        col_user=constants.DEFAULT_USER_COL,
+        col_item=constants.DEFAULT_ITEM_COL,
+        col_score=constants.DEFAULT_PREDICTION_COL
+    ):
+        """Initialize model parameters
+
+        Args:
+            col_user (str): user column name
+            col_item (str): item column name
+            col_score (str): score column name
+        """
+        self.col_item = col_item
+        self.col_user = col_user
+        self.col_score = col_score
+
+        self._model_df = None
+        self._item_frequencies = None
+
+        # set flag to disallow calling fit() before predict()
+        self._is_fit = False
+
+    def __repr__(self):
+        """Make a friendly, human-readable object summary string."""
+        return (
+            f"{self.__class__.__name__}(user_colname={self.col_user}, "
+            f"item_colname={self.col_item}, "
+            f"score_colname={self.col_score} "
+        )
+
+    def _check_dataframe(self, df,
+                         expected_columns=None):
+        """Verify input is a data frame and has expected columns.
+
+        Checks if the input is a DataFrame object and if so,
+        if the mandatory columns are present. Also checks for
+        duplicate rows.
+        """
+        if not isinstance(df, pd.DataFrame):
+            raise TypeError((
+                "Input argument must be a pandas DataFrame. "
+                f"Instead got {type(df)}!"))
+
+        if not expected_columns:
+            expected_columns = [self.col_user, self.col_item]
+        for col in expected_columns:
+            if col not in df.columns:
+                raise KeyError((
+                    f"Column {col} not found in DataFrame!"))
+
+        # check there are no duplicate rows in dataframe
+        if df.groupby(expected_columns).size().max() > 1:
+            raise ValueError("Duplicate rows found!!")
+
+    def fit(self, df):
+        """Fit the FBT recommender using an input DataFrame.
+
+        Fitting of model involves computing a item-item co-occurrence
+        matrix: how many people bought watched a pair of movies?
+
+        Args:
+            df (pd.DataFrame): DataFrame of users and items
+
+        Returns:
```

Comment:
There is no returned value, so no "Returns" is needed in the doc string.

Reply:
resolved.